remote_fetch.go

```diff
@@ -80,33 +80,72 @@ func decompressResponse(bodyBytes []byte, contentLength int64, cacheLocator Cach
 		return errorFetchingPostgres(err)
 	}
 
+	cacheLocation, _ := cacheLocator()
+
+	if err := os.MkdirAll(filepath.Dir(cacheLocation), 0755); err != nil {
+		return errorExtractingPostgres(err)
+	}
+
 	for _, file := range zipReader.File {
 		if !file.FileHeader.FileInfo().IsDir() && strings.HasSuffix(file.FileHeader.Name, ".txz") {
-			archiveReader, err := file.Open()
-			if err != nil {
-				return errorExtractingPostgres(err)
+			if err := decompressSingleFile(file, cacheLocation); err != nil {
+				return err
 			}
 
-			archiveBytes, err := io.ReadAll(archiveReader)
-			if err != nil {
-				return errorExtractingPostgres(err)
-			}
+			// we have successfully found the file, return early
+			return nil
+		}
+	}
 
-			cacheLocation, _ := cacheLocator()
+	return fmt.Errorf("error fetching postgres: cannot find binary in archive retrieved from %s", downloadURL)
+}
 
-			if err := os.MkdirAll(filepath.Dir(cacheLocation), 0755); err != nil {
-				return errorExtractingPostgres(err)
-			}
+func decompressSingleFile(file *zip.File, cacheLocation string) error {
+	renamed := false
 
-			if err := os.WriteFile(cacheLocation, archiveBytes, file.FileHeader.Mode()); err != nil {
-				return errorExtractingPostgres(err)
-			}
+	archiveReader, err := file.Open()
+	if err != nil {
+		return errorExtractingPostgres(err)
+	}
 
-			return nil
+	archiveBytes, err := io.ReadAll(archiveReader)
+	if err != nil {
+		return errorExtractingPostgres(err)
+	}
+
+	// if multiple processes attempt to extract
+	// to prevent file corruption when multiple processes attempt to extract at the same time
+	// first to a cache location, and then move the file into place.
+	tmp, err := os.CreateTemp(filepath.Dir(cacheLocation), "temp_")
+	if err != nil {
+		return errorExtractingPostgres(err)
+	}
+	defer func() {
+		// if anything failed before the rename then the temporary file should be cleaned up.
+		// if the rename was successful then there is no temporary file to remove.
+		if !renamed {
+			if err := os.Remove(tmp.Name()); err != nil {
+				panic(err)
+			}
 		}
+	}()
+
+	if err := os.WriteFile(tmp.Name(), archiveBytes, file.FileHeader.Mode()); err != nil {
```

Comment:
This is a bit weird, because `tmp` created above is already an open file. So you have two open descriptors to the same "name", and below when you close `tmp` I'm not sure what the end state is. You should probably call `tmp.Close()` before this line.

Overall I'm a bit confused about all the code you're adding... to me only the changes in this file are needed, so I haven't even looked at the rest.